src/main/java/org/prebid/server/analytics/reporter/greenbids/GreenbidsAnalyticsReporter.java

```diff
@@ -0,0 +1,323 @@
+
+        if (imps == null || imps.isEmpty()) {
+            throw new IllegalArgumentException("imps is null or empty");
+        }
+
+        final Long auctionTimestamp = Optional.of(auctionContext.getBidRequest())
+                .map(BidRequest::getExt)
+                .map(ExtRequest::getPrebid)
+                .map(ExtRequestPrebid::getAuctiontimestamp)
+                .orElse(null);
+
+        final long auctionElapsed = auctionTimestamp != null ? System.currentTimeMillis() - auctionTimestamp : 0L;
+
+        final Map<String, Bid> seatsWithBids = Optional.ofNullable(bidResponse.getSeatbid())
+                .orElseGet(Collections::emptyList)
+                .stream()
+                .filter(seatBid -> !seatBid.getBid().isEmpty())
+                .collect(
+                        Collectors.toMap(
+                                SeatBid::getSeat,
+                                seatBid -> seatBid.getBid().get(0),
+                                (existing, replacement) -> existing
+                        )
+                );
+
+        final List<SeatNonBid> seatNonBids = auctionContext.getBidRejectionTrackers().entrySet().stream()
+                .map(entry -> toSeatNonBid(entry.getKey(), entry.getValue()))
+                .filter(seatNonBid -> !seatNonBid.getNonBid().isEmpty())
+                .toList();
+
+        final Map<String, NonBid> seatsWithNonBids = Optional.of(seatNonBids)
+                .orElseGet(Collections::emptyList)
+                .stream()
+                .filter(seatNonBid -> !seatNonBid.getNonBid().isEmpty())
+                .collect(
+                        Collectors.toMap(
+                                SeatNonBid::getSeat,
+                                seatNonBid -> seatNonBid.getNonBid().get(0),
+                                (existing, replacement) -> existing
+                        )
+                );
+
+        final CommonMessage commonMessage = createCommonMessage(auctionContext, auctionElapsed);
+
+        return addBidResponseToMessage(
+                commonMessage,
+                imps,
+                seatsWithBids,
+                seatsWithNonBids
+        );
+    }
+
+    @Override
+    public <T> Future<Void> processEvent(T event) {
+        AuctionContext greenbidsAuctionContext = null;
+        BidResponse greenbidsBidResponse = null;
+
+        if (event instanceof AmpEvent ampEvent) {
+            greenbidsAuctionContext = ampEvent.getAuctionContext();
+            greenbidsBidResponse = ampEvent.getBidResponse();
+        } else if (event instanceof AuctionEvent auctionEvent) {
+            greenbidsAuctionContext = auctionEvent.getAuctionContext();
+            greenbidsBidResponse = auctionEvent.getBidResponse();
+        }
+        assert greenbidsBidResponse != null && greenbidsAuctionContext != null;
+
+        this.greenbidsId = UUID.randomUUID().toString();
+        this.billingId = UUID.randomUUID().toString();
```

Comment:
self-review: in Prebid.js `greenbidsId` and `billingId` are generated UUID inside the [GreenbidsRTDAdapter](https://github.com/prebid/Prebid.js/blob/3350dee8ee32e07941d7871caf66b75bd66dc93b/modules/greenbidsRtdProvider.js#L4). At the moment before the PBS GreenbidsRTDAdapter is done we will generate it inside PBS GreenbidsAnalyticsReporter